batch/batch/worker/worker.py

```diff
@@ -1793,8 +1803,16 @@ async def run(self):
                         await self.cleanup()
                     finally:
                         _, exc, _ = sys.exc_info()
+                        # mark_complete moves ownership of `mjs_fut` into another task
+                        # but if it either is not run or is cancelled then we need
+                        # to cancel MJS
                         if not isinstance(exc, asyncio.CancelledError):
-                            await self.mark_complete()
+                            try:
+                                await self.mark_complete(mjs_fut)
+                            except asyncio.CancelledError:
+                                mjs_fut.cancel()
+                        else:
+                            mjs_fut.cancel()
```

Comment:
Your change preserves semantics, so I'm totally good with it. However, it's worth noting that cancellation muffles exceptions. I think that's right in this case because we got interrupted so it doesn't matter what happened to MJS.